tests/malice1.rs

```diff
@@ -0,0 +1,158 @@
+///
+/// The Makers identify the situation and gets their money back via contract txs. This is
+/// a potential DOS on Makers. But Taker would loose money too for doing this.
+#[tokio::test]
+async fn malice1_taker_broadcast_contract_prematurely() {
+    // ---- Setup ----
+
+    let makers_config_map = [
+        (6102, MakerBehavior::Normal),
+        (16102, MakerBehavior::Normal),
+    ];
+
+    // Initiate test framework, Makers.
+    // Taker has normal behavior.
+    let (test_framework, taker, makers) = TestFramework::init(
+        None,
+        makers_config_map.into(),
+        Some(TakerBehavior::BroadcastContractAfterFullSetup),
+    )
+    .await;
+
+    // Fund the Taker and Makers with 3 utxos of 0.05 btc each.
+    for _ in 0..3 {
+        let taker_address = taker
+            .write()
+            .unwrap()
+            .get_wallet_mut()
+            .get_next_external_address()
+            .unwrap();
+        test_framework.send_to_address(&taker_address, Amount::from_btc(0.05).unwrap());
+        makers.iter().for_each(|maker| {
+            let maker_addrs = maker
+                .get_wallet()
+                .write()
+                .unwrap()
+                .get_next_external_address()
+                .unwrap();
+            test_framework.send_to_address(&maker_addrs, Amount::from_btc(0.05).unwrap());
+        })
+    }
+
+    // confirm balances
+    test_framework.generate_1_block();
+
+    let org_maker_balances = makers
+        .iter()
+        .map(|maker| {
+            maker
+                .get_wallet()
+                .read()
+                .unwrap()
+                .balance(false, false)
+                .unwrap()
+        })
+        .collect::<BTreeSet<_>>();
+
+    let org_take_balance = taker
+        .read()
+        .unwrap()
+        .get_wallet()
+        .balance(false, false)
+        .unwrap();
+
+    // ---- Start Servers and attempt Swap ----
+
+    // Start the Maker server threads
+    let maker_threads = makers
+        .iter()
+        .map(|maker| {
+            let maker_clone = maker.clone();
+            let thread = thread::spawn(move || {
+                start_maker_server(maker_clone).unwrap();
+            });
+            thread
+        })
+        .collect::<Vec<_>>();
+
+    // Start swap
+    thread::sleep(Duration::from_secs(20)); // Take a delay because Makers take time to fully setup.
+    let swap_params = SwapParams {
+        send_amount: 500000,
+        maker_count: 2,
+        tx_count: 3,
+        required_confirms: 1,
+        fee_rate: 1000,
+    };
+
+    // Spawn a Taker coinswap thread.
+    let taker_clone = taker.clone();
+    let taker_thread = thread::spawn(move || {
+        taker_clone
+            .write()
+            .unwrap()
+            .send_coinswap(swap_params)
+            .unwrap();
+    });
+
+    // Wait for Taker swap thread to conclude.
+    taker_thread.join().unwrap();
```

Comment:
Could we consider using `taker_thread.join()` instead of `taker_thread.join().unwrap()`? In the event of a thread panic, we could implement error handling using a `match` statement.

Reply:
Yes we can and we should and assert the error with a match, for the cases where Taker would through hard errors. 

In this particular case it won't, so its okay to unwrap here.